src/lib.rs

```diff
@@ -85,11 +80,15 @@ enum Icon {
     Pencil,
 }
 
-/// HUD Actions
+/// Actions
 ///
-/// Events within the HUD that are returned back to the user.
-pub enum HUDAction {
-    
+/// Events that are returned back to the user.
+pub enum Action {
+    HUD,
+    Window,
+    Menu,
+    Button,
+    Highlight,
```

Comment:
I'm curious what event `Highlight` is?  I think `Hud` and `Menu` will end up being the same action (I'd add a comment to the `Menu` variant saying that it's from the HUD).  I'm guessing you want this PR to also close https://github.com/ardaku/sprengui/issues/10 as well, in which case we're missing a few variants:

 - `Search`
 - `SelectTab`
 - `MoveTab`
 - `Entry` (which I'm imagining would include `Button`, can add an additional enum `EntryAction` with the `Button` variant, a `Field` variant, a `Slider` variant, etc.)

Would you mind adding comments to each variant explaining what it is?  I'm also starting to think we might want a separate type for each variant (using associated data) doing something like this:

```
/// ID for a menu action
#[derive(Copy, Clone, PartialEq, Eq)]
pub struct MenuAction {
    action_id: usize,
    menu_id: usize,
}

pub enum Action {
    /// A menu action produced by the HUD
    Menu(MenuAction),
    ...
}
```

If you don't know what to put for fields for each action type you can just do:

```rust
pub struct MenuAction;
```

Or just guess.  Since this is the initial version I expect the definitions to change over time, so it's not super necessary to get each action type right.

Reply:
I think it might be better to get rid of the `HUD` variant in favor of `Menu`, in order to avoid using acronyms for variant names.

Reply:
will each variant have specific action ids attached to them so we know the different ways a user can interact with each element? i.e. button might have actions for hover, and pressed. while slider might have hover, grab(bool) and move. also where will we keep the data such elements produce? for example where the slider is currently, and the text within a field.

Reply:
A slider would probably only need to send and action when moved, where it would just return an `f32`; Grab and hover would events would probably only be used internally for animations.  There's also "flattening" we could try, so variants would be named `FieldText`, `FieldNumber`, etc. within the `Action` enum.

Reply:
hmm in that case do you think we should also have just two or three generic action structs based on the general type of the object? I guess I'm trying to just get a sense of overall how the data will flow. I'm gonna whip up an ms paint diagram to see if I can't organize the way I'm seeing it

Reply:
I think the way you a have it is good, I wouldn't change the `Action` enum as it is.  I'll look at it closer later today for specific events.  I think all the `Field` events should be `EntryAction` events.

Reply:
![image](https://user-images.githubusercontent.com/94191498/178794427-97b24d89-0b5f-4c31-822e-f83eeee76303.png)
how is this?

Reply:
We shouldn't need an event queue in this crate.  The window crate send events via its own `Notifier`, which will be translated to sprengui `Action`s in the `Notifier` impl.  The `Vec` of elements (which would be called widgets), looks correct to me.  Although, there should be a `Vec` of widgets for each `Page`, and the `Window` will have a `Vec<Page>` or something like that.  We might also want to add another action event for switching `Page` that takes an index.  `Page` will have `Button` and `Slider`, etc. (doesn't contain Window).  Window will also contain the HUD menu, tabs, page switcher, headerbar, etc.  Hopefully that's not too confusing.

Reply:
Hmm are we sure we want to include a button and a slider by default? wouldn't it be better to leave it up to the developer? or maybe just include a slider that appears/disappears dynamically based on the size of the page vs. the size of the window? Or we could try and set up the layout based on window size so that elements automatically adjust their position relative to the size of the window. I imagine each window would have a default of 1 page and then the developer can add pages if they want? I suppose lower level widgets like buttons and fields might have parent pages which are associated with a specific window. Let me know if I'm putting the cart before the horse here.

Reply:
I think we'll want to plan to have them, just because there are many types of apps that would require such controls and it's better if people don't have to reinvent the wheel for each app.  A page could have any number of buttons or sliders, and I'm also thinking buttons will also work better as "button groups".  A slider or button group would always take up 100% width (helps with making protable to mobile).  Button groups might max out at 8 buttons.  Pages are really designed for form-like input, while canvas is used for most everything else (so some apps may have zero pages).  Pages would be linear, and there will possibly be child activities as a separate concept (I'm not sure how much of that will be built into sprengui yet, it might be left to the user).

Reply:
> I imagine sliders will be invisible on mobile?

Sliders are just a thin horizontal box with a drag-able handle that goes linearly from 0 to 1, so I don't see why it wouldn't work on mobile.

> Also I was thinking we could include some templates in a folder so that devs dont have to start from scratch and then keep the rest more or less completely modular. Devs could then create their own templates if they choose.

Yeah, we'll have an examples directory showing off different types of program interfaces (also so we can have some nice screenshots).

Reply:
> Sliders are just a thin horizontal box with a drag-able handle that goes linearly from 0 to 1, so I don't see why it wouldn't work on mobile

Yes but that style of interaction typically feels unnatural on mobile. It would be more natural/useful if the slider was invisible, and we capture a drag input that then adjusts the current position of the handle on the slider behind the scenes. Otherwise the app will feel like a desktop application running on a mobile device which I think defeats the purpose.

Reply:
I always thought sliders were a pretty natural touchscreen widget.  Whether you're on mobile or desktop they can be imprecise, but also useful for quickly selecting colors in a color picker (although maybe a color wheel is even better in that situation).  We can probably just forget about sliders.

Reply:
I agree, lets worry about them later. For the time being, is there anything in the latest PR that you would like altered/developed further before merging?

Reply:
I'll check again in about 3 hours, if that's ok.  I'm guessing it's good enough to merge as-is.

Reply:
Just because there are some things we don't know how they're going to work yet anyway.